File: backend/app/routes/testimonials.py
# backend/app/routes/testimonials.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity

from ..extensions import db
from ..models import Testimonial, Product, User

testimonials_bp = Blueprint('testimonials', __name__)

def is_admin():
    user_id = get_jwt_identity()
    user = User.query.get(user_id)
    return user and user.is_admin()

@testimonials_bp.route('/', methods=['GET'])
def get_testimonials():
    """Get approved testimonials"""
    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 10, type=int), 50)
    product_id = request.args.get('product_id', type=int)
    
    query = Testimonial.query.filter_by(is_approved=True)
    
    if product_id:
        query = query.filter_by(product_id=product_id)
    
    pagination = query.order_by(Testimonial.created_at.desc())\
        .paginate(page=page, per_page=per_page, error_out=False)
    
    return jsonify({
        'success': True,
        'data': {
            'testimonials': [t.to_dict() for t in pagination.items],
            'pagination': {
                'page': pagination.page,
                'pages': pagination.pages,
                'total': pagination.total,
                'per_page': pagination.per_page
            }
        }
    }), 200

@testimonials_bp.route('/', methods=['POST'])
@jwt_required()
def add_testimonial():
    """Add a testimonial/review"""
    user_id = get_jwt_identity()
    data = request.get_json()
    
    # Validate required fields
    if not data.get('product_id'):
        return jsonify({'success': False, 'message': 'Product ID required'}), 400
    
    if not data.get('rating') or data['rating'] < 1 or data['rating'] > 5:
        return jsonify({'success': False, 'message': 'Rating must be between 1 and 5'}), 400
    
    if not data.get('review'):
        return jsonify({'success': False, 'message': 'Review text required'}), 400
    
    # Check if user already reviewed this product
    existing = Testimonial.query.filter_by(
        user_id=user_id,
        product_id=data['product_id']
    ).first()
    
    if existing:
        return jsonify({'success': False, 'message': 'You have already reviewed this product'}), 400
    
    # Check if user purchased this product
    from ..models import Order, OrderItem
    has_purchased = db.session.query(OrderItem)\
        .join(Order)\
        .filter(
            Order.user_id == user_id,
            OrderItem.product_id == data['product_id'],
            Order.status == 'delivered'
        ).first() is not None
    
    testimonial = Testimonial(
        user_id=user_id,
        product_id=data['product_id'],
        rating=data['rating'],
        title=data.get('title'),
        review=data['review'],
        is_approved=has_purchased  # Auto-approve if purchased
    )
    
    db.session.add(testimonial)
    db.session.commit()
    
    # Update product rating
    product = Product.query.get(data['product_id'])
    if product:
        product.update_rating()
    
    return jsonify({
        'success': True,
        'message': 'Review submitted successfully',
        'data': testimonial.to_dict()
    }), 201

@testimonials_bp.route('/<int:testimonial_id>', methods=['PUT'])
@jwt_required()
def update_testimonial(testimonial_id):
    """Update own testimonial"""
    user_id = get_jwt_identity()
    
    testimonial = Testimonial.query.get(testimonial_id)
    
    if not testimonial:
        return jsonify({'success': False, 'message': 'Review not found'}), 404
    
    if testimonial.user_id != user_id:
        return jsonify({'success': False, 'message': 'Unauthorized'}), 403
    
    data = request.get_json()
    
    if data.get('rating'):
        testimonial.rating = data['rating']
    if data.get('title'):
        testimonial.title = data['title']
    if data.get('review'):
        testimonial.review = data['review']
    
    # Reset approval status
    testimonial.is_approved = False
    testimonial.approved_at = None
    
    db.session.commit()
    
    # Update product rating
    if testimonial.product_id:
        product = Product.query.get(testimonial.product_id)
        if product:
            product.update_rating()
    
    return jsonify({
        'success': True,
        'message': 'Review updated',
        'data': testimonial.to_dict()
    }), 200

@testimonials_bp.route('/<int:testimonial_id>', methods=['DELETE'])
@jwt_required()
def delete_testimonial(testimonial_id):
    """Delete own testimonial"""
    user_id = get_jwt_identity()
    user = User.query.get(user_id)
    
    testimonial = Testimonial.query.get(testimonial_id)
    
    if not testimonial:
        return jsonify({'success': False, 'message': 'Review not found'}), 404
    
    if testimonial.user_id != user_id and not user.is_admin():
        return jsonify({'success': False, 'message': 'Unauthorized'}), 403
    
    product_id = testimonial.product_id
    
    db.session.delete(testimonial)
    db.session.commit()
    
    # Update product rating
    if product_id:
        product = Product.query.get(product_id)
        if product:
            product.update_rating()
    
    return jsonify({'success': True, 'message': 'Review deleted'}), 200

@testimonials_bp.route('/user/my-reviews', methods=['GET'])
@jwt_required()
def get_my_reviews():
    """Get current user's reviews"""
    user_id = get_jwt_identity()
    
    reviews = Testimonial.query.filter_by(user_id=user_id)\
        .order_by(Testimonial.created_at.desc())\
        .all()
    
    return jsonify({
        'success': True,
        'data': [review.to_dict() for review in reviews]
    }), 200

# Admin routes
@testimonials_bp.route('/admin/testimonials', methods=['GET'])
@jwt_required()
def admin_get_testimonials():
    """Get all testimonials (admin only)"""
    if not is_admin():
        return jsonify({'success': False, 'message': 'Admin access required'}), 403
    
    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 20, type=int), 100)
    status = request.args.get('status')  # 'approved', 'pending'
    
    query = Testimonial.query
    
    if status == 'approved':
        query = query.filter_by(is_approved=True)
    elif status == 'pending':
        query = query.filter_by(is_approved=False)
    
    pagination = query.order_by(Testimonial.created_at.desc())\
        .paginate(page=page, per_page=per_page, error_out=False)
    
    return jsonify({
        'success': True,
        'data': {
            'testimonials': [t.to_dict() for t in pagination.items],
            'pagination': {
                'page': pagination.page,
                'pages': pagination.pages,
                'total': pagination.total,
                'per_page': pagination.per_page
            }
        }
    }), 200

@testimonials_bp.route('/admin/testimonials/<int:testimonial_id>/approve', methods=['POST'])
@jwt_required()
def approve_testimonial(testimonial_id):
    """Approve testimonial (admin only)"""
    if not is_admin():
        return jsonify({'success': False, 'message': 'Admin access required'}), 403
    
    testimonial = Testimonial.query.get(testimonial_id)
    
    if not testimonial:
        return jsonify({'success': False, 'message': 'Review not found'}), 404
    
    data = request.get_json()
    testimonial.approve(admin_response=data.get('response'))
    
    return jsonify({
        'success': True,
        'message': 'Review approved',
        'data': testimonial.to_dict()
    }), 200

@testimonials_bp.route('/admin/testimonials/<int:testimonial_id>/reject', methods=['POST'])
@jwt_required()
def reject_testimonial(testimonial_id):
    """Reject testimonial (admin only)"""
    if not is_admin():
        return jsonify({'success': False, 'message': 'Admin access required'}), 403
    
    testimonial = Testimonial.query.get(testimonial_id)
    
    if not testimonial:
        return jsonify({'success': False, 'message': 'Review not found'}), 404
    
    testimonial.reject()
    
    return jsonify({'success': True, 'message': 'Review rejected'}), 200

File: frontend/src/services/testimonials.js
// src/services/testimonials.js
import { get, post, put, del } from './api';

// Get all approved testimonials (public)
export const getTestimonials = async (limit = 10, page = 1) => {
  try {
    const response = await get(`/testimonials?limit=${limit}&page=${page}`);
    
    if (response.data.success) {
      return {
        success: true,
        testimonials: response.data.data.testimonials,
        pagination: response.data.data.pagination,
      };
    }
    return { success: false, testimonials: [], pagination: null };
  } catch (error) {
    console.error('Get testimonials error:', error);
    return { success: false, testimonials: [] };
  }
};

// Get product testimonials
export const getProductTestimonials = async (productId, limit = 5) => {
  try {
    const response = await get(`/products/${productId}/testimonials?limit=${limit}`);
    
    if (response.data.success) {
      return { success: true, testimonials: response.data.data };
    }
    return { success: false, testimonials: [] };
  } catch (error) {
    console.error('Get product testimonials error:', error);
    return { success: false, testimonials: [] };
  }
};

// Add testimonial (authenticated users)
export const addTestimonial = async (testimonialData) => {
  try {
    const response = await post('/testimonials', {
      product_id: testimonialData.productId,
      rating: testimonialData.rating,
      review: testimonialData.review,
    });
    
    if (response.data.success) {
      toast.success('Thank you for your review!');
      return { success: true, testimonial: response.data.data };
    }
    return { success: false, error: response.data.message };
  } catch (error) {
    console.error('Add testimonial error:', error);
    const message = error.response?.data?.message || 'Failed to submit review';
    toast.error(message);
    return { success: false, error: message };
  }
};

// Update testimonial (owner only)
export const updateTestimonial = async (testimonialId, data) => {
  try {
    const response = await put(`/testimonials/${testimonialId}`, {
      rating: data.rating,
      review: data.review,
    });
    
    if (response.data.success) {
      toast.success('Review updated');
      return { success: true, testimonial: response.data.data };
    }
    return { success: false, error: response.data.message };
  } catch (error) {
    console.error('Update testimonial error:', error);
    return { success: false, error: error.message };
  }
};

// Delete testimonial (owner or admin)
export const deleteTestimonial = async (testimonialId) => {
  try {
    const response = await del(`/testimonials/${testimonialId}`);
    
    if (response.data.success) {
      toast.success('Review deleted');
      return { success: true };
    }
    return { success: false, error: response.data.message };
  } catch (error) {
    console.error('Delete testimonial error:', error);
    return { success: false, error: error.message };
  }
};

// Get user's testimonials
export const getUserTestimonials = async () => {
  try {
    const response = await get('/user/testimonials');
    
    if (response.data.success) {
      return { success: true, testimonials: response.data.data };
    }
    return { success: false, testimonials: [] };
  } catch (error) {
    console.error('Get user testimonials error:', error);
    return { success: false, testimonials: [] };
  }
};

// Admin: Get all testimonials (including pending)
export const getAllTestimonials = async (filters = {}, page = 1, limit = 20) => {
  try {
    const params = new URLSearchParams();
    if (filters.status) params.append('status', filters.status);
    if (filters.search) params.append('search', filters.search);
    params.append('page', page);
    params.append('limit', limit);
    
    const response = await get(`/admin/testimonials?${params.toString()}`);
    
    if (response.data.success) {
      return {
        success: true,
        testimonials: response.data.data.testimonials,
        pagination: response.data.data.pagination,
      };
    }
    return { success: false, testimonials: [], pagination: null };
  } catch (error) {
    console.error('Get all testimonials error:', error);
    return { success: false, testimonials: [] };
  }
};

// Admin: Approve testimonial
export const approveTestimonial = async (testimonialId) => {
  try {
    const response = await post(`/admin/testimonials/${testimonialId}/approve`);
    
    if (response.data.success) {
      toast.success('Review approved');
      return { success: true };
    }
    return { success: false, error: response.data.message };
  } catch (error) {
    console.error('Approve testimonial error:', error);
    return { success: false, error: error.message };
  }
};

// Admin: Reject testimonial
export const rejectTestimonial = async (testimonialId, reason = '') => {
  try {
    const response = await post(`/admin/testimonials/${testimonialId}/reject`, { reason });
    
    if (response.data.success) {
      toast.success('Review rejected');
      return { success: true };
    }
    return { success: false, error: response.data.message };
  } catch (error) {
    console.error('Reject testimonial error:', error);
    return { success: false, error: error.message };
  }
};

// Get testimonial statistics (admin)
export const getTestimonialStats = async () => {
  try {
    const response = await get('/admin/testimonials/stats');
    
    if (response.data.success) {
      return { success: true, stats: response.data.data };
    }
    return { success: false, stats: null };
  } catch (error) {
    console.error('Get testimonial stats error:', error);
    return { success: false, stats: null };
  }
};